taskcluster/mozillavpn_taskgraph/actions/release_promotion.py

```diff
@@ -0,0 +1,147 @@
+                "minimum": 1,
+                "title": "The release build number",
+                "description": (
+                    "The release build number. Starts at 1 per "
+                    "release version, and increments on rebuild."
+                ),
+            },
+            "do_not_optimize": {
+                "type": "array",
+                "description": (
+                    "Optional: a list of labels to avoid optimizing out "
+                    "of the graph (to force a rerun)."
+                ),
+                "items": {
+                    "type": "string",
+                },
+            },
+            "revision": {
+                "type": "string",
+                "title": "Optional: revision to ship",
+                "description": ("Optional: the revision to ship."),
+            },
+            "release_promotion_flavor": {
+                "type": "string",
+                "description": "The flavor of release promotion to perform.",
+                "default": "promote-addons",
+                "enum": sorted(graph_config["release-promotion"]["flavors"].keys()),
+            },
+            "rebuild_kinds": {
+                "type": "array",
+                "description": (
+                    "Optional: an array of kinds to ignore from the previous "
+                    "graph(s)."
+                ),
+                "items": {
+                    "type": "string",
+                },
+            },
+            "previous_graph_ids": {
+                "type": "array",
+                "description": (
+                    "Optional: an array of taskIds of decision or action "
+                    "tasks from the previous graph(s) to use to populate "
+                    "our `previous_graph_kinds`."
+                ),
+                "items": {
+                    "type": "string",
+                },
+            },
+        },
+        "required": [
+            "release_promotion_flavor",
+            "build_number",
+        ],
+    },
+)
+def release_promotion_action(parameters, graph_config, input, task_group_id, task_id):
+    release_promotion_flavor = input["release_promotion_flavor"]
+    promotion_config = graph_config["release-promotion"]["flavors"][
+        release_promotion_flavor
+    ]
+
+    target_tasks_method = promotion_config["target-tasks-method"].format(
+        project=parameters["project"]
+    )
+    rebuild_kinds = input.get("rebuild_kinds") or promotion_config.get(
+        "rebuild-kinds", []
+    )
+    do_not_optimize = input.get("do_not_optimize") or promotion_config.get(
+        "do-not-optimize", []
+    )
+
+    # make parameters read-write
+    parameters = dict(parameters)
+    # Build previous_graph_ids from ``previous_graph_ids`` or ``revision``.
+    previous_graph_ids = input.get("previous_graph_ids")
+    if not previous_graph_ids:
+        previous_graph_ids = [find_decision_task(parameters, graph_config)]
```

Comment:
I also wonder if `previous_graph_ids`, `do-not-optimize` and `rebuild-kinds` will be useful or not.. But they likely fall into the "doesn't hurt to have them" category.

Reply:
I'll leave them in, doesn't hurt to have them!